File: python/word_search.py
class Solution(object):
    # O(m * 4^n) time | O(n) space
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """

        def backtrack(r, c, i):
            if i == len(word):
                return True
            
            if (r < 0 or r >= len(board) or c < 0 or c >= len(board[r]) 
                    or word[i] != board[r][c] or board[r][c] == '#'):
                return False

            curr = board[r][c]    
            board[r][c] = '#'

            res = (backtrack(r + 1, c, i + 1) or backtrack(r - 1, c, i + 1) 
                   or backtrack(r, c + 1, i + 1) or backtrack(r, c - 1, i + 1))

            board[r][c] = curr
            return res


        for r in range(len(board)):
            for c in range(len(board[r])):
                if board[r][c] == word[0]:
                    if backtrack(r, c, 0):
                        return True
        
        return False
    
solution = Solution()

board = [
    ['A', 'B', 'C', 'E'],
    ['S', 'F', 'C', 'S'],
    ['A', 'D', 'E', 'E']
]

word = "ABCCED"

print(solution.exist(board, word))    

File: src/main/java/january_2026/WordSearch.java
package january_2026;

import java.util.HashMap;
import java.util.Map;

public class WordSearch {

    public static void main(String[] args) {
//        char[][] board = {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
//        String word = "ABCB";

        char[][] board = {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
        String word = "ABCCED";

        WordSearch wordSearch = new WordSearch();
        boolean exist = wordSearch.exist(board, word);
        System.out.println(exist);
    }

    // O(n * m * 4^l) time | O(l) space
    public boolean exist(char[][] board, String word) {

        for (int r = 0; r < board.length; r++) {
            for (int c = 0; c < board[r].length; c++) {
                if (word.charAt(0) == board[r][c]) {
                    boolean match = backtrack(board, r, c, 0, word);
                    if (match) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private boolean backtrack(char[][] board, int r, int c, int idx, String word) {
        if (idx == word.length()) {
            return true;
        }
        if (r < 0 || r >= board.length || c < 0 || c >= board[r].length || board[r][c] == '#' || word.charAt(idx) != board[r][c]) {
            return false;
        }
        char curr = board[r][c];
        board[r][c] = '#';
        int[][] directions = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        for (int[] dir : directions) {
            boolean match = backtrack(board, r + dir[0], c + dir[1], idx + 1, word);
            if (match) {
                return true;
            }
        }
        board[r][c] = curr;
        return false;
    }


}
